blog/_posts/2017-01-24-cudanative-nvidia.md

```diff
@@ -0,0 +1,234 @@
+`Pkg.add(...)` away!
+
+What's missing from this list is the lowest abstraction level, where you can create kernels
+and manage execution like you would in CUDA. This is where CUDAnative.jl comes in.
+
+
+
+## Native GPU programming with CUDAnative.jl
+
+[CUDAnative.jl](https://github.com/JuliaGPU/CUDAnative.jl) is a brand-new package adding
+native GPU programming capabilities to the Julia programming language. Together with a
+package like [CUDAdrv.jl](https://github.com/JuliaGPU/CUDAdrv.jl) or
+[CUDArt.jl](https://github.com/JuliaGPU/CUDArt.jl) for interfacing with the CUDA libraries,
+it is now possible to do CUDA development in Julia without the need for an external language
+or compiler. As an introductory example, the following listing shows how to compute the sum
+of two vectors:
+
+```julia
+using CUDAdrv, CUDAnative
+
+function kernel_vadd(a, b, c)
+    i = (blockIdx().x-1) * blockDim().x + threadIdx().x
+    c[i] = a[i] + b[i]
+
+    return nothing
+end
+
+dev = CuDevice(0)
+ctx = CuContext(dev)
+
+# generate some data
+len = 512
+a = rand(Int, len)
+b = rand(Int, len)
+
+# allocate & upload on the GPU
+d_a = CuArray(a)
+d_b = CuArray(b)
+d_c = similar(d_a)
+
+# execute and fetch results
+@cuda (1,len) kernel_vadd(d_a, d_b, d_c)
+c = Array(d_c)
+
+destroy(ctx)
+```
+
+The real workhorse of this example is the `@cuda` macro, which generates specialized code
+for compiling the kernel function to GPU assembly, uploading it to the driver, and preparing
+the execution environment. Together with Julia's JIT compiler, this results in a very
+efficient kernel launch sequence, avoiding the runtime overhead typically associated with
+dynamic languages.
+
+<!-- This seems a bit short. Maybe elaborate more? -->
+
+<!-- Mention libNVVM? -->
+
+Behind the scenes, CUDAnative.jl hooks into the Julia compiler to lower kernel functions to
+LLVM IR, an intermediate form of code used by the LLVM compiler framework. We've created
+[LLVM.jl](https://github.com/maleadt/LLVM.jl) to interface with the framework, and use it to
+compile IR to PTX code, the virtual instruction set used to program CUDA GPUs. Finally, we
+use CUDAdrv.jl to interact with the CUDA driver: upload the PTX code, compile it to
+hardware-specific SASS, and launch the kernel. <!-- TODO: See image 1 for an overview. -->
+
+![Overview of the different components making up the native programming
+support.](/images/blog/2017-01-24-cudanative/overview.png)
+
+<!-- Overview picture: show GPU, SASS, CUDA? -->
+
+By reusing the Julia compiler to generate IR we avoid multiple language implementations with
+slightly different semantics. That means you can use all features of the Julia language,
+like metaprogramming or dynamically-typed multimethods, to implement GPU functions. There
+are still some limitations, for example we haven't ported Julia's runtime library or
+implemented a GPU garbage-collector yet, but these aren't structural limitations.
```

Comment:
Can you be more explicit about the kinds of restrictions? For example, I suspect I can't call a julia function in the middle of a kernel.